common/src/main/scala/explore/utils/package.scala

```diff
@@ -170,3 +176,66 @@ extension (bytes: NonNegLong)
 
 object IsExpanded extends NewType[Boolean]
 type IsExpanded = IsExpanded.Type
+
+extension [F[_], O](s: Stream[F, O])
+  /**
+   * Combine items within a given duration using the given function.
+   *
+   * When an item is received in the stream, a timer is started. Within that time, items are
+   * combined until the timer expires, at which point the combined item is emitted.
+   *
+   * If no items are received within the given duration, no item is emitted.
+   */
+  def reduceWithin(d: FiniteDuration, f: (O, O) => O)(using F: Temporal[F]): Stream[F, O] =
+    given Semigroup[O] with
+      def combine(x: O, y: O): O = f(x, y)
+    reduceSemigroupWithin(d)
+
+  /**
+   * Combine items within a given duration using Semigroup.
+   *
+   * When an item is received in the stream, a timer is started. Within that time, items are
+   * combined until the timer expires, at which point the combined item is emitted.
+   *
+   * If no items are received within the given duration, no item is emitted.
+   */
+  def reduceSemigroupWithin(
```

Comment:
Sorry I didn't get around to reviewing earlier.

Can the implementation be simplified by using `groupWithin`?

Reply:
Unfortunately no, the implementation of groupWithin always immediately outputs 1 element after the given duration has passed, instead of waiting for more, and it's not customizable